extensions/2.0/Khronos/KHR_materials_variants/README.md

```diff
@@ -0,0 +1,131 @@
+
+![](examples/shoes/photo.png)
+
+## Design Goals
+
+_This section is non-normative._
+
+This extension aims specifically to give asset creators and consumers a way to communicate _static configurability_.  In other words, we explicitly optimize for the case of representing asset variants which are:
+
+* **finite** - the list of variants is bounded; multi-dimensional variants (e.g. color + texture) are possible but must ultimately be representable as a flattened list
+* **premade** - the variants are produced at asset authoring time
+* **holistic** - while composable, the variants are dictated wholesale at the granularity of the glTF asset; sub-variants or hierarchical variants are not considered
+
+In digital commerce, each variant may correspond to a distinct SKU or stocked product code, which naturally meet these requirements.  In games or movies, each variant may represent a fixed set of curated asset skins which can be represented, moved across the wire, and loaded efficiently.
+
+A non-goal of this extension is to serve _configuration authorship_ use cases (i.e. configuration builders or "configurators"), which necessarily entail a separate set of concerns, are more subject to complex application-specific business logic, and may benefit from dynamic and continuous configurability and complex builder rules.
+
+While not designed around this use case, we note however that this extension does implicitly communicate at the primitive level which materials are "valid" for that primitive, for any such applications which may be interested in leveraging this information.
+
+## Variants
+
+We introduce a simple variant tagging extension scheme, that allows for high-level runtime swapping between which `material` is used to shade a given `mesh primitive`: the extension root contains a mandatory `mapping` property, which is an array of objects, each one associating some set of variants with a material reference.
+
+Imagine a sneaker with shoelace holes that are made from materials that depend on the overall shoe colour in non-obvious ways:
+
+| Variants                           | Material                        |
+| ---------------------------------- | ------------------------------- |
+| `sneaker_yellow`, `sneaker_orange` | `shoelace_hole_material_brown`  |
+| `sneaker_red`                      | `shoelace_hole_material_purple` |
+| `sneaker_black`                    | `shoelace_hole_material_yellow` |
+
+(_the authors of this spec are not product designers, apologies for the dubious colour choices_)
+
+Application-specific logic defines, for a given instance of a glTF asset, a single active variant.  The currently active `material` for a `mesh primitive` is found by stepping through this array of mappings in order, and selecting the first one which contains somewhere in its list of variants the currently active variant. The corresponding material is assigned to the mesh.  If none match, fall back on vanilla glTF behaviour.
```

Comment:
If only a single variant may be active, then the order of the mappings shouldn't matter, right? If a variant appears twice in the mappings list, the later of the two can never be selected.

Reply:
Yeah fair enough.  Technically the schema itself does allow for these never-selected mappings to exist, so the order piece just makes it unambiguous which mapping wins - the first should always be selected.  The alternative would be to just explicate/strictly require that for each primitive the variants mentioned in each mapping object must be locally unique -- and a validator for this extension should fail in these cases.. do you think that is preferable?

Reply:
If the spec will state normatively that only a single variant may be active, then I think it should also say normatively that variants cannot appear twice in the same primitive mappings list. And probably also that the 'variants' array cannot be empty for any mapping?

It sounds like a material index _could_ twice in the mapping, e.g. associated with two different names?

Reply:
Makes sense to me.  Will add this

Re: material index, yes though this would be semantically identical to including those different names in the same mapping item.  If we will explicitly prohibit duplicating variants and empty mappings, it probably makes sense to disallow this as well so that there is only one obvious way to do things

Reply:
The fact that a mapping is the only place you can attach names/extras/extensions is arguably some justification for allowing duplicate material indices, but see https://github.com/KhronosGroup/glTF/pull/1681#issuecomment-667424806 for more on that.